File: 0085-maximal-rectangle/0085-maximal-rectangle.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalRectangle = function(matrix) {
    if(!matrix.length){
        return 0;
    }
     let n =  matrix.length;
    let m = matrix[0].length;
    let prefixSum = Array.from({ length: n }, () => Array(m).fill(0));
   
    let maxArea = 0;
    for(let j = 0;j<m;j++){
        let sum = 0;
        for(let i = 0;i<n;i++){

            sum+=1;
            if(matrix[i][j]==="0"){
                sum = 0; 
            }
            prefixSum[i][j] = sum
        }

    }

    for(let i = 0;i<n;i++){
        maxArea = Math.max(maxArea,largestRectangleArea(prefixSum[i]))
    }
    return maxArea;
};


var largestRectangleArea = function(heights) {
    let nsl = getNsl(heights);
    let nsr = getNsr(heights);
    let finalArea = 0;
    for(let i = 0;i<heights.length;i++){
        let width = nsr[i]-nsl[i] -1;
        let newArea = heights[i]*width;
        finalArea = Math.max(finalArea,newArea);
    }
    return finalArea;
    
};

var getNsl = function(heights){
    let st = [];
    let res = [];
    for(let i =0;i < heights.length;i++){
        while(st.length!==0 && heights[st[st.length-1]]>= heights[i]){
            st.pop();
        }
        res[i] = st.length === 0 ? -1:st[st.length - 1];
        st.push(i);
    } 
    return res;
}

var getNsr = function(heights){
    let st = [];
    let res = [];
    for(let i = heights.length - 1;i>=0;i--){
        while(st.length!==0 && heights[st[st.length-1]]> heights[i]){
            st.pop();
        }
       res[i] = st.length === 0 ? heights.length:st[st.length - 1];
        st.push(i);
    } 
    return res;
}

File: 0085-maximal-rectangle/0085-maximal-rectangle.py
class Solution:
    def getNsl(self,heights):
        n = len(heights)
        stack = []
        res = [0]*n
        for i in range(n):
            while stack and heights[stack[-1]] >= heights[i]:
                stack.pop()
            res[i] = -1 if not stack else stack[-1]
            stack.append(i)
        return res

    def getNsr(self,heights):
        n = len(heights)
        stack = []
        res = [0]*n
        for i in range(n-1,-1,-1):
            while stack and heights[stack[-1]] > heights[i]:
                stack.pop()
            res[i] = n if not stack else stack[-1]
            stack.append(i)
        return res

    def largestRectangleArea(self, heights: List[int]) -> int:
        nsl = self.getNsl(heights)
        nsr = self.getNsr(heights)
        n = len(heights)
        finalArea = 0
        for i in range(n):
            width = nsr[i] - nsl[i] - 1
            area = heights[i]*width
            finalArea = max(area,finalArea)
        return finalArea



    def maximalRectangle(self, matrix: List[List[str]]) -> int:
        n = len(matrix)
        m = len(matrix[0])
        maxArea = 0
        prefixSum = [[0 for _ in range(m)] for _ in range(n)]
        for j in range(m):
            newSum = 0
            for i in range(n):
                newSum+=1
                if matrix[i][j] =="0":
                    newSum = 0
                
                prefixSum[i][j] = newSum
        
        for i in range(n):
            maxArea = max(maxArea,self.largestRectangleArea(prefixSum[i]))

        return maxArea

                


        